File: year2019/day02/java/Solution.java
import java.nio.file.*;
import java.util.*;

public class Solution {
    private static int[] CODES;

    public static void main(String[] args) { 
        try {
            setup(Files.readString(Paths.get("../puzzle.txt")));
            System.out.println(part1());
            setup(Files.readString(Paths.get("../puzzle.txt")));
            System.out.println(part2());
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
    
	public static int part1() {
        return simulate(CODES, 12, 2);
    }
    
    public static int part2() {
        for (int noun = 0; noun < CODES.length; noun++) {
            for (int verb = 0; verb < CODES.length; verb++) {
                if (simulate(Arrays.copyOf(CODES, CODES.length), noun, verb) == 19690720) {
                    return 100 * noun + verb;
                }
            }
        }
        return -1;
    }

    public static void setup(String input) {
        CODES = Arrays.stream(input.split(",")).mapToInt(Integer::parseInt).toArray();
    }

    private static int simulate(int[] codes, int noun, int verb) {
        codes[1] = noun;
        codes[2] = verb;
        for(int i = 0; codes[i] != 99; i += 4) {
            if (codes[i] == 1) {
                codes[codes[i + 3]] = codes[codes[i + 1]] + codes[codes[i + 2]];
            } else {
                codes[codes[i + 3]] = codes[codes[i + 1]] * codes[codes[i + 2]];
            }
        }
        return codes[0];
    }
}

File: year2019/day02/python/solution.py
def main():
  print(part_2())
  print(part_1())


def part_1():
  return simulate(input, 12, 2)


def part_2():
  input.pop()
  for noun in range(len(input)):
    for verb in range(len(input)):
      if simulate(input[:], noun, verb) == 19690720:
        return 100 * noun + verb


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [int(num) for num in file.read().split(',')]


def simulate(codes, noun, verb):
  codes[1], codes[2] = noun, verb
  i = 0
  while codes[i] != 99:
    codes[codes[i + 3]] = codes[codes[i + 1]] + codes[codes[i + 2]] if codes[i] == 1 else codes[codes[i + 1]] * codes[codes[i + 2]]
    i += 4

  return codes[0]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()